File: retest.ts


import * as core from '@actions/core'
import * as github from '@actions/github'
import axios from 'axios'
import { OctokitResponse } from "@octokit/types";
import { GithubReactionType, CheckRunsType, CreateReactionType, Retest, RetestResult, PR, Env } from './types'

function cachedProperty(_: unknown, key: string, descriptor: PropertyDescriptor): PropertyDescriptor {
  const originalGetter = descriptor.get
  if (!originalGetter) {
    throw new Error('The decorated property must have a getter.')
  }


  const cachedValueKey = Symbol(`__cached_${key}`)

  descriptor.get = function (this: any): any {

    if (!this[cachedValueKey]) {
      this[cachedValueKey] = originalGetter.call(this)
    }

    return this[cachedValueKey]
  }

  return descriptor
}

class RetestCommand {

  public checks: CheckRunsType['data']['check_runs']
  public env: Env
  public name = ''
  public pr: PR

  constructor(env: Env, pr: PR, checks: CheckRunsType['data']['check_runs']) {
    this.env = env
    this.checks = checks
    this.pr = pr
  }

  retest = async (): Promise<RetestResult> => {
    if (!this.env) {
      core.warning(`Failed parsing env`)
      return {errors: 0, retested: 0}
    }
    const retestables = await this.getRetestables()
    if (Object.keys(retestables).length === 0) {
      return {errors: 0, retested: 0}
    }
    return await this.retestRuns(retestables)
  }

  retestExternal = async (check: Retest): Promise<number> => {
    if (check.method == 'patch') {
      try {
        await axios.patch(check.url, {}, check.config)
        /* eslint-disable  prettier/prettier */
      } catch (error: any) {
        if (!axios.isAxiosError(error) || !error.response) {
          core.error('No response received')
          return 1
        }
        core.error(`External API call failed: ${check.url}`)
        core.error(error.response.data.message)
        return 1
      }
    }
    return 0
  }

  retestOctokit = async (pr: PR, check: Retest): Promise<number> => {

    const method = check.method || 'POST'
    const rerunURL = `${method} ${check.url}`
    if (rerunURL.endsWith('rerun-failed-jobs')) {
      console.log(`Retesting failed job (pr #${pr.number}): ${check.name}`)
    } else {
      console.log(`Restarting check (pr #${pr.number}): ${check.name}`)
    }
    const rerunResponse = await this.env.octokit.request(rerunURL, check.config || {})
    if ([200, 201].includes(rerunResponse.status)) {
      if (rerunURL.endsWith('rerun-failed-jobs')) {
        process.stdout.write(`::notice::Retry success: (${check.name})\n`)
      } else {
        process.stdout.write(`::notice::Check restarted: (${check.name})\n  ${rerunResponse.data.html_url}\n`)
      }
      return 0
    } else {
      if (rerunURL.endsWith('rerun-failed-jobs')) {
        core.error(`Retry failed: (${check.name}) ... ${rerunResponse.status}`)
      } else {
        core.error(`Failed restarting check: ${rerunResponse.status}`)
      }
      return 1
    }
  }

  retestRuns = async (retestables: Array<Retest>): Promise<RetestResult> => {

    let errors = 0
    for (const check of retestables) {
      if (!check.octokit) {
        errors += await this.retestExternal(check)
      } else {
        errors += await this.retestOctokit(this.pr, check)
      }
    }
    return {retested: Object.keys(retestables).length, errors: errors}
  }

  getRetestables = async (): Promise<Array<Retest>> => {
    return []
  }
}

class GithubRetestCommand extends RetestCommand {

  name = 'Github'

  constructor(env: Env, pr: PR, checks: CheckRunsType['data']['check_runs']) {
    super(env, pr, checks)
  }

  getRetestables = async (): Promise<Array<Retest>> => {
    const failedChecks: any[] = []
    const checks = this.checks.filter((checkRun) => {
      return checkRun.app?.slug
    })
    checks.forEach(async (check: any) => {
      if (check.conclusion !== 'failure' && check.conclusion !== 'cancelled') {
        return
      }
      if (this.env.debug) {
        console.log(
          `Check ${check.conclusion}: ${check.name}\n\n  ${check.html_url}\n\n`
          + `  https://github.com/${this.env.owner}/${this.env.repo}/actions/runs/${check.external_id}\n`)
      }
      // run failed jobs.
      failedChecks.push({
        name: check.name || 'unknown',
        url: `/repos/${this.env.owner}/${this.env.repo}/actions/runs/${check.external_id}/rerun-failed-jobs`,
        octokit: true,
      })
      // TODO: Update the old check to mention restart
      // Create a new check from the old
      const toDelete = ['pull_requests', 'app', 'check_suite', 'conclusion', 'node_id', 'started_at', 'completed_at', 'id']
      Object.keys(check).forEach((key) => {
        if (key.startsWith("url") || key.endsWith("url") || toDelete.includes(key)) {
          delete check[key]
        }
      })
      Object.keys(check.output).forEach((key) => {
        if (key.startsWith("annotation")) {
          delete check.output[key]
        }
      })

      check.output.title = check.output.title.replace('failure', 'restarted')
      const lines = check.output.text.split('\n')
      const line0 = lines[0].replace('Check run finished (failure :x:)', 'Check run restarted')
      check.output.text = `${line0}\n${lines.slice(1).join('\n')}`
      check.output.summary = 'Check is running again'
      check.status = 'in_progress'

      failedChecks.push({
        name: check.name || 'unknown',
        url: `/repos/${this.env.owner}/${this.env.repo}/check-runs`,
        octokit: true,
        config: {data: check},
      })
    })
    return failedChecks
  }
}

class RetestCommands {

  @cachedProperty
  get env(): Env {
    const token = core.getInput('token') || process.env['GITHUB_TOKEN']
    if (!token || token === '') throw new TypeError('`token` must be set')

    const pr = core.getInput('pr-url')
    const comment = parseInt(core.getInput('comment-id'))
    console.log(token)
    const octokit = github.getOctokit(token)
    // Create the octokit client
    const nwo = process.env['GITHUB_REPOSITORY'] || '/'
    const [owner, repo] = nwo.split('/')
    const debug = Boolean(process.env.CI_DEBUG && process.env.CI_DEBUG != 'false')

    return {
      debug,
      token,
      octokit,
      nwo,
      owner,
      repo,
      comment,
      pr,
    }
  }

  checks = async (pr: PR): Promise<CheckRunsType['data']['check_runs']> => {
    const response: CheckRunsType = await this.env.octokit.rest.checks.listForRef({
      owner: this.env.owner,
      per_page: 100,
      repo: this.env.repo,
      ref: pr.commit,
      filter: 'latest',
    })
    const checks = response.data.check_runs
    if (!checks) return []
    if (this.env.debug) {
      checks.forEach((checkRun) => {
        console.log(`Found check (${checkRun.id}/${checkRun.app?.slug}/${checkRun.conclusion || 'incomplete'}): ${checkRun.name}`);
      });
    }
    return checks
  }

  getPR = async (): Promise<PR | void> => {
    if (!this.env.pr || !this.env.pr) {
      return
    }
    const response: OctokitResponse<any> = await this.env.octokit.request(this.env.pr)
    const data = response.data
    if (!data) return
    if (this.env.debug) {
      console.log(`Running /retest command for PR #${data.number}`)
      console.log(`PR branch: ${data.head.ref}`)
      console.log(`Latest PR commit: ${data.head.sha}`)
    }
    return {
      number: data.number,
      branch: data.head.ref,
      commit: data.head.sha,
    }
  }

  retesters = async (): Promise<Array<RetestCommand>> => {
    if (!this.env) {
      return []
    }
    const pr = await this.getPR()
    if (!pr) {
      return []
    }
    const checks = await this.checks(pr)
    const retesters: Array<RetestCommand> = [new GithubRetestCommand(this.env, pr, checks)]
    return retesters
  }

  addReaction = async (reaction: GithubReactionType = 'rocket'): Promise<void> => {
    const addReactionResponse: CreateReactionType['response'] = await this.env.octokit.rest.reactions.createForIssueComment({
      owner: this.env.owner,
      repo: this.env.repo,
      comment_id: this.env.comment,
      content: reaction,
    })
    if ([200, 201].includes(addReactionResponse.status)) {
      console.log(`Reacted to comment ${reaction}`)
    } else {
      core.error(`Failed reacting to comment ${reaction}`)
    }
  }

  retest = async (): Promise<void> => {
    const result: RetestResult = {errors: 0, retested: 0}
    for (const retester of await this.retesters()) {
      const retested = await retester.retest()
      result.errors += retested.errors
      result.retested += retested.retested
    }
    if (result.errors !== 0) {
      await this.addReaction('-1')
    }
    if (result.retested === 0) {
      await this.addReaction('confused')
    } else {
      await this.addReaction()
    }
  }
}

const run = async (): Promise<void> => {
  try {
    const retesters = new RetestCommands()
    await retesters.retest()
  } catch (error) {
    if (error instanceof Error) {
      console.error(error.message)
    }
    core.setFailed(`retest-action failure: ${error}`)
  }
}

// Don't auto-execute in the test environment
if (process.env['NODE_ENV'] !== 'test') {

  run()
}

export default run

File: retest/retest.go
package retest

import (
	"context"
	"fmt"
	"log"
	"net/http"
	"os"
	"strconv"
	"strings"

	"github.com/actions-go/toolkit/core"
	"github.com/actions-go/toolkit/github"
	github2 "github.com/google/go-github/v42/github"
)

var (
	githubClient = github.NewClient()
)

func InitRetestCommands() *Runtime {

	commentInput, _ := core.GetInput("comment-id")
	comment, _ := strconv.Atoi(commentInput)
	pr, _ := core.GetInput("pr-url")
	nwo := os.Getenv("GITHUB_REPOSITORY")
	debug := os.Getenv("CI_DEBUG") != "" && os.Getenv("CI_DEBUG") != "false"
	var repo, owner string

	if nwo != "" {
		owner = strings.Split(nwo, "/")[0]
		repo = strings.Split(nwo, "/")[1]
	} else {
		log.Fatal("GITHUB_REPOSITORY must not be nil")
	}

	return &Runtime{
		Pr:      pr,
		Comment: comment,
		Repo:    repo,
		Nwo:     nwo,
		Owner:   owner,
		Debug:   debug,
	}

}

func getSuffix(pr string) int {

	prSplit := strings.Split(pr, "/")
	prNumber, _ := strconv.Atoi(prSplit[len(prSplit)-1])

	return prNumber
}

func getPR(rt *Runtime) *PullRequest {

	if rt.Pr == "" {

		log.Fatal("env.pr url is nil")
	}

	pr, prResp, err := githubClient.PullRequests.Get(
		context.Background(),
		rt.Owner,
		rt.Repo,
		getSuffix(rt.Pr),
	)

	if pr == nil && prResp.StatusCode != 200 && prResp.StatusCode != 201 && err != nil {

		log.Fatal("pr not found, err: ", err)
	}

	return &PullRequest{
		Branch: pr.Head.GetRef(),
		Number: pr.GetNumber(),
		Commit: pr.Head.GetSHA(),
	}

}

func addReaction(rt *Runtime, content string) bool {

	_, response, err := githubClient.Reactions.CreateIssueCommentReaction(
		context.Background(),
		rt.Owner,
		rt.Repo,
		int64(rt.Comment),
		content,
	)

	if response.StatusCode != 200 && response.StatusCode != 201 && err != nil {

		log.Fatal("failed to add reaction, error: ", err)
		return false
	}

	return true

}

func getFailedJos(rt *Runtime, pr *PullRequest) (failedChecks []*GHRetest) {

	ref, response, err := githubClient.Checks.ListCheckRunsForRef(
		context.Background(),
		rt.Owner,
		rt.Repo,
		pr.Commit,
		nil,
	)

	if response.StatusCode != 200 && response.StatusCode != 201 && err != nil {

		log.Fatal("failed to get check runs, error: ", err)

	}

	for order, run := range ref.CheckRuns {

		if rt.Debug {
			log.Printf("check run: %v, order: %v\n", run, order)
		}

		if run.GetConclusion() == "failure" {

			failedChecks = append(failedChecks, &GHRetest{
				Name: run.GetName(),
				Url:  run.GetDetailsURL(),
			})
		}

	}

	return failedChecks
}

func rerunJobs(rt *Runtime, failedJobs []*GHRetest) (result *GHRetestResult) {

	var errorNum int

	for _, job := range failedJobs {

		fmt.Printf("retesting check: %v\n %v\n", job.Name, job.Url)

		u := fmt.Sprintf("repos/%v/%v/actions/jobs/%v/rerun", rt.Owner, rt.Repo, getSuffix(job.Url))
		req, err := githubClient.NewRequest(http.MethodPost, u, nil)

		if err != nil {
			log.Fatal("failed to create request, error: ", err)
			return nil
		}

		job := new(github2.WorkflowJob)
		resp, err := githubClient.Do(context.Background(), req, job)

		if resp.StatusCode != 201 && err != nil {

			errorNum++
			log.Fatal("failed to retest job, error: ", err)
			return nil
		}

	}

	return &GHRetestResult{
		Error:    errorNum,
		Retested: len(failedJobs),
	}
}

func retest() {

	rt := InitRetestCommands()
	pr := getPR(rt)
	failedJosList := getFailedJos(rt, pr)

	if len(failedJosList) == 0 {

		log.Println("no failed checks found")
		return
	}

	if rt.Debug {
		log.Printf("Runtime info: %v\n: ", rt)
		log.Printf("pr info: %v", pr)
	}

	result := rerunJobs(rt, failedJosList)
	if result.Error != 0 {

		addReaction(rt, "-1")
	}
	if result.Error == 0 {

		log.Println("all checks have been restarted")
		addReaction(rt, "rocket")
	} else {

		log.Printf("failed to restart some checks, error times: %v\n", result.Error)
		addReaction(rt, "confused")
	}

}

func Run() {

	defer func() {
		if err := recover(); err != nil {

			log.Println("retest error: ", err)
			core.SetFailedf("Retest action failure, error is %v\n", err)
		}
	}()

	retest()
}
